{{cookiecutter.profile_name}}/lsf_submit.py

```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/env python3
+"""
+lsf-submit.py
+
+Script to wrap bsub sync command for Snakemake. Uses the following job or
+cluster parameters:
+
++ `threads`
++ `resources`
+    - `mem_mb`: Expected memory requirements in megabytes. Overrides
+      cluster.mem_mb
++ `cluster`
+    - `mem_mb`: Memory that will be requested for the cluster for the job.
+      Overriden by resources.mem_mb, if present.
+      `resources`
+    - `queue`: Which queue to run job on
+    - `logdir`: Where to log stdout/stderr from cluster command
+    - `output`: Name of stdout logfile
+    - `error`: Name of stderr logfile
+    - `jobname`: Job name (with wildcards)
+
+Author: Michael B Hall
+Adapted from: https://github.com/jaicher/snakemake-sync-bq-sub
+"""
+
+import sys
+import subprocess
+from pathlib import Path
+from snakemake.utils import read_job_properties
+from typing import List
+import re
+if not __name__.startswith("tests.src."):
+    sys.path.append(str(Path(__file__).parent.absolute()))
+    from OSLayer import OSLayer
+    from CookieCutter import CookieCutter
+else:
+    from .OSLayer import OSLayer
+    from .CookieCutter import CookieCutter
+
+
+class LSF_Submit:
+    def __init__(self, argv: List[str]):
+        self._jobscript = argv[-1]
+        self._cluster_cmd = " ".join(argv[1:-1])
+        self._job_properties = read_job_properties(self._jobscript)
+        self.random_string = OSLayer.get_random_alphanumerical_string()
```

Comment:
can you explain this? I don't like the idea of adding a random string to log file names.

Reply:
additionally, using https://docs.python.org/3/library/uuid.html is probably a more robust way to do this.

Reply:
The purpose of this random string is if a pipeline is executed several times (due to some jobs failing, and then several resumes), different jobs can get the same id in different executions of the pipeline.
For example, let's say that in execution 1 of the pipeline, job pandora_compare with id=10 fails, while job pandora_index with id=5 succeeds. Some changes are made, and in execution 2, pandora_compare gets assigned id=5. In this case, if bjobs fail, we could inadvertently read that the job with id=5 (i.e. pandora_compare) succeeded, while in fact it was the previous pandora_index. This random 64 alphanumerical string is to avoid these conflicts of different jobs with same ids in different reexecutions of the pipeline. In case you are very unlucky of having the conflict(different jobs with same job id and the same random 64-length alphanumerical string in a reexecution of the pipeline), we actually remove the previous log (https://github.com/leoisl/snakemake-lsf/blob/44b0ab5bffc3561c211571ea6a9511e798564161/%7B%7Bcookiecutter.profile_name%7D%7D/lsf_submit.py#L180) in order for us to not get confused that the job ran fine when it failed. However, the goal is to keep as much log as we want, we don't want to delete log files, hence the 64-char random string.

Do you propose using https://docs.python.org/3/library/uuid.html#uuid.uuid4 instead of generating a random n-length string? I don't actually understand uuid4 that well, I don't manage to understand what are its benefits over a random string (I am assuming uuid4 generates a random string based on current time of the seed? But we can't actually control its length nor how it is made?).

Reply:
It's a Universally Unique IDentifier (UUID).  
We don't control it, but that's kind of the point I guess. `uuid4()` returns a random UUID (string length 36)  
> [A UUID is 128 bits long, and can guarantee uniqueness across space and time.][1]

[This SO Q/A](https://stackoverflow.com/questions/52740862/what-is-the-purpose-of-a-uuid-or-guid-compared-to-a-random-string) also answers your question somewhat. And [this one](https://stackoverflow.com/a/39776/5299417) talks about the possibility of collisions.

[1]: https://tools.ietf.org/html/rfc4122.html

Reply:
Yeah, that is actually what we need. Implemented in https://github.com/leoisl/snakemake-lsf/commit/aedfa2b09f4b254c91618cf08634178e60c5fec4